ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Node.hs

```diff
@@ -564,48 +564,58 @@ instance CardanoHardForkConstraints c
   ProtocolInfo
 -------------------------------------------------------------------------------}
 
--- | Parameters needed to transition from one era to the next.
-type ProtocolTransitionParams :: Type -> Type -> Type
-data family ProtocolTransitionParams x y
-
--- | Parameters needed to transition from Byron to Shelley.
-data instance ProtocolTransitionParams ByronBlock (ShelleyBlock proto era) =
-    ProtocolTransitionParamsByronToShelley {
-        transitionByronToShelleyTranslationContext :: Core.TranslationContext era
-      , transitionByronToShelleyTrigger            :: TriggerHardFork
-      }
+newtype CardanoHardForkTriggers = CardanoHardForkTriggers {
+    getCardanoHardForkTriggers :: NP (K TriggerHardFork) (CardanoShelleyEras StandardCrypto)
+  }
 
--- | Parameters needed to transition from a Shelley era to the next.
-data instance ProtocolTransitionParams (ShelleyBlock proto era) (ShelleyBlock proto' era') =
-    ProtocolTransitionParamsIntraShelley {
-        transitionIntraShelleyTranslationContext :: Core.TranslationContext era'
-      , transitionIntraShelleyTrigger            :: TriggerHardFork
-      }
+pattern CardanoHardForkTriggers' ::
```

Comment:
In general I find the `'` problematic when it comes to constructor because it's hard to remember which one is which. But I don't have a good suggestion for an alternative so feel free to ignore this. (Perhaps sth like `MkCardanoHardForkTriggers`?

Reply:
Why not use a classic Haskell98 record type here? There is nothing special about `TriggerHardFork` at the type level. This would simplify initialization of protocolInfo in db-analyzer and cardano-api 10 fold.
Does is it really bring any more type safety when used using generics sop approach?
In my understanding of this whole setup this does not protect me at all from reading the wrong thing from json config file and mixing up the order.

Reply:
> This would simplify initialization of protocolInfo in db-analyzer and cardano-api 10 fold.

What would get simpler? From a usage perspective, exposing this pattern synonym means that it is exactly as convenient as a plain old record type, users don't have to be aware of the SOP stuff at all. See eg the integration in [cardano-api](https://github.com/input-output-hk/cardano-api/pull/270/files#diff-501ffc6173492d93fcdd4a112b067f1bffb106547a6ca798187c84012599be09R839-R847).

---

Using an `NP` to represent this allows us to know statically what era a trigger is for, which means we can eg use `eraProtVerLow` to figure out the trigger versions automatically, as is done eg in db-analyser since this PR:
https://github.com/input-output-hk/ouroboros-consensus/blob/29c591b9eb6cd2c2bdb803ad3c63cdda7f6f3ed1/ouroboros-consensus-cardano/src/unstable-cardano-tools/Cardano/Tools/DBAnalyser/Block/Cardano.hs#L226

This means that in the future, people might not even have to specify the trigger versions (unless they want to override them). Also see #319 for broader goals in making `protocolInfoCardano` simpler to use.

Reply:
Fully agreed that the second one is much clearer for someone without sop-core/Consensus experience and should defnitely be preferred in contexts where that is not a given (eg in node, that is also why I did not use this style in eg cardano-api as linked above); but eg one can't rely on types to catch mistakes like accidentally swapping `@Alonzo` and `@Allegra` (don't want to say that that is a huge deal here).

But more generally, the changes in this PR and not requiring anyone to write code like in the first approach (see the cardano-api integration), they can just use the convenient pattern synonym, so I do not see any barrier. We definitely want to make invoking `protocolInfoCardano` as simple as possible to use (#319), and I think in this PR is a step in that direction (could remove quite a few now-redundant lines in the `protocolInfoCardano` invocation in cardano-api).